openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -1,144 +1,203 @@
         return super(CallStackMixin, self).save(*args, **kwargs)
 
     def delete(self, *args, **kwargs):
         """
-        Logs before delete and overrides respective model API delete()
+        Logs before delete() and overrides respective model API delete()
         """
-        capture_call_stack(type(self))
+        if hasattr(self, 'model'):
+            capture_call_stack(self.model)
+        else:
+            capture_call_stack(type(self))
         return super(CallStackMixin, self).delete(*args, **kwargs)
 
 
 class CallStackManager(Manager):
-    """ A Manager class which overrides the default Manager class for getting call stacks
+    """ Manager class which overrides the default Manager class for getting call stacks
     """
     def get_query_set(self):
-        """overriding the default queryset API method
+        """ Override the default queryset API method
         """
-        capture_call_stack(self.model)
+        if hasattr(self, 'model'):
+            capture_call_stack(self.model)
+        else:
+            capture_call_stack(type(self))
         return super(CallStackManager, self).get_query_set()
 
 
-def donottrack(*classes_not_to_be_tracked):
-    """function decorator which deals with toggling call stack
-    Args:
-        classes_not_to_be_tracked: model classes where tracking is undesirable
+def donottrack(*entities_not_to_be_tracked):
+    """ Decorator which halts tracking for some entities for specific functions
+
+    Arguments:
+        entities_not_to_be_tracked: entities which are not to be tracked
+
     Returns:
         wrapped function
     """
+    @wrapt.decorator
+    def real_donottrack(wrapped, instance, args, kwargs):  # pylint: disable=W0613
+        """ Takes function to be decorated and returns wrapped function
 
-    def real_donottrack(function):
-        """takes function to be decorated and returns wrapped function
+        Arguments:
+            wrapped - The wrapped function which in turns needs to be called by wrapper function.
+            instance - The object to which the wrapped function was bound when it was called.
+            args - The list of positional arguments supplied when the decorated function was called.
+            kwargs - The dictionary of keyword arguments supplied when the decorated function was called.
 
-        Args:
-            function - wrapped function i.e. real_donottrack
         """
-        def wrapper(*args, **kwargs):
-            """ wrapper function for decorated function
-            Returns:
-                wrapper function i.e. wrapper
-            """
-            if len(classes_not_to_be_tracked) == 0:
-                global TRACK_FLAG  # pylint: disable=W0603
-                current_flag = TRACK_FLAG
-                TRACK_FLAG = False
-                function(*args, **kwargs)
-                TRACK_FLAG = current_flag
+        if entities_not_to_be_tracked:
+            global HALT_TRACKING  # pylint: disable=W0603
```

Comment:
Likewise here. You don't need this `global` statement. Also, you should use the full name of pylint errors, not the codes.